File: VB/Client/XpoHelper.vb
Imports Microsoft.VisualBasic
Imports DXSample.BO
Imports DevExpress.Xpo
Imports DXSample.DataStore
Imports System.Configuration
Imports DevExpress.Xpo.Metadata

Namespace DXSample.Client
	Public NotInheritable Class XpoHelper
		Private Sub New()
		End Sub
		Public Shared Function GetNewSession() As Session
			Return New Session(DataLayer)
		End Function

		Public Shared Function GetNewUnitOfWork() As UnitOfWork
			Return New UnitOfWork(DataLayer)
		End Function

		Private ReadOnly Shared lockObject As Object = New Object()

'INSTANT VB TODO TASK: There is no VB.NET equivalent to 'volatile':
'ORIGINAL LINE: static volatile IDataLayer fDataLayer;
		Private Shared fDataLayer As IDataLayer
		Private Shared ReadOnly Property DataLayer() As IDataLayer
			Get
				If fDataLayer Is Nothing Then
					SyncLock lockObject
						If fDataLayer Is Nothing Then
							fDataLayer = GetDataLayer()
						End If
					End SyncLock
				End If
				Return fDataLayer
			End Get
		End Property

		Private Shared Function GetDataLayer() As IDataLayer
			XpoDefault.Session = Nothing
			Dim dict As XPDictionary = New ReflectionDictionary()
			dict.GetDataStoreSchema(GetType(Customer).Assembly)
			Return New ThreadSafeDataLayer(dict, New CommandChannelServiceDataStore(ConfigurationManager.ConnectionStrings("XpoServiceUrl").ConnectionString))
		End Function
	End Class
End Namespace

File: CS/Client/XpoHelper.cs
using DXSample.BO;
using DevExpress.Xpo;
using DXSample.DataStore;
using System.Configuration;
using DevExpress.Xpo.Metadata;

namespace DXSample.Client {
    public static class XpoHelper {
        public static Session GetNewSession() {
            return new Session(DataLayer);
        }

        public static UnitOfWork GetNewUnitOfWork() {
            return new UnitOfWork(DataLayer);
        }

        private readonly static object lockObject = new object();

        static volatile IDataLayer fDataLayer;
        static IDataLayer DataLayer {
            get {
                if (fDataLayer == null) {
                    lock (lockObject) {
                        if (fDataLayer == null) {
                            fDataLayer = GetDataLayer();
                        }
                    }
                }
                return fDataLayer;
            }
        }

        private static IDataLayer GetDataLayer() {
            XpoDefault.Session = null;
            XPDictionary dict = new ReflectionDictionary();
            dict.GetDataStoreSchema(typeof(Customer).Assembly);
            return new ThreadSafeDataLayer(dict, 
                new CommandChannelServiceDataStore(ConfigurationManager.ConnectionStrings["XpoServiceUrl"].ConnectionString));
        }
    }
}